[1.9] Granting modders access to the villager trades and fixes issues with villager names.

thilokru opened the pull request:
With this PR I intend to grant modders the ability to add trades to already existing villagers and to allow them to create new kinds of villagers easily. Also, with the new Event, modders can manipulate trades more easily and flexibly.

Usecases:

- Modifying a villager careers trade options (-> VillagerCareer methods)
- Modifying the villager trades when they are created (-> Event)
- Creating custom careers to already existing professions
- Allows modders, if they are adding new careers to vanilla professions, to set custom names in cases where vanilla knows only one career per profession

All the above mentioned features have a reference implementation in the example mod.

williewillus commented on src/main/java/net/minecraftforge/event/entity/living/VillagerTradePopulationEvent.java:
```diff
@@ -0,0 +1,21 @@
+    public final VillagerCareer career;
+    public final List<ITradeList> trades;
+    public final int skillLevel;
+    
```
change to private and add getters

Also, javadocs

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -151,33 +208,161 @@ private void init()
         VillagerProfession prof = new VillagerProfession("minecraft:farmer", "minecraft:textures/entity/villager/farmer.png");
         {
             register(prof, 0);
-            (new VillagerCareer(prof, "farmer")).init(VanillaTrades.trades[0][0]);
```
This entire section: No
The original arrays are in the format below because they are more easily diffed against the original arrays during MC updates to make sure we did not miss anything. The init functions and trades arrays are kept private because modders should NOT be accessing them .

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -136,10 +138,65 @@ private void register(VillagerProfession prof, int id)
     private FMLControlledNamespacedRegistry<VillagerProfession> professions = PersistentRegistryManager.createRegistry(PROFESSIONS, VillagerProfession.class, null, 0, 1024, true, null, null, null);
-    public IForgeRegistry<VillagerProfession> getRegistry() { return this.professions; }
 
+    public IForgeRegistry<VillagerProfession> getRegistry()
```
Fix all your reformatting spam it's cluttering the PR -.-

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -136,10 +138,65 @@ private void register(VillagerProfession prof, int id)
+     * @return the {@link VillagerProfession}, or, if not available, null
+     */
+    public VillagerProfession getVillagerProfession(String name)
+    {
```
No, they can access the registry itself and pull it out no need for bouncer methods.

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -136,10 +138,65 @@ private void register(VillagerProfession prof, int id)
+     *            the trade that will be registered.
+     */
+    public void registerTrade(VillagerCareer career, int careerLevel, ITradeList trade)
+    {
```
Trades are to be managed by the Career object itself NOT a static method here.

LexManos commented on patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch:
```diff
@@ -120,16 +120,26 @@
          int i = this.field_175563_bv - 1;
          int j = this.field_175562_bw - 1;
 -        EntityVillager.ITradeList[][] aentityvillager$itradelist1 = aentityvillager$itradelist[i];
-+        EntityVillager.ITradeList[][] aentityvillager$itradelist1 = this.getProfessionForge().getCareer(i).getTrades();
```
The hook was designed to return a dual array to prevent the patch from being large. So ya, don't do that.

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -200,9 +385,14 @@ private void register(VillagerCareer career)
             Validate.isTrue(career.profession == this, "Attempted to register career for the wrong profession.");
             career.id = careers.size();
             careers.add(career);
+            TRADES.put(career, new ArrayList<List<ITradeList>>());
```
Having second lists associated with careers we control is just dumb. That's what the trade list is in the career object.

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -200,9 +385,14 @@ private void register(VillagerCareer career)
+            TRADES.put(career, new ArrayList<List<ITradeList>>());
+        }
+
+        public ResourceLocation getSkin()
```
Again more formatting spam

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -212,6 +402,27 @@ public VillagerCareer getCareer(int id)
             return this.careers.get(0);
         }
+        
+        public String getName(){
```
No, bouncer methods again bad! Registry objects have a standard method to get the name we want modders to get used to using it.

LexManos commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -237,16 +447,26 @@ public String getName()
+            boolean validLevel = skillLevel >= 0 && skillLevel < skilledTradeList.size();
+            List<ITradeList> trades = Lists.newArrayList(skilledTradeList.get(validLevel ? skillLevel : 0));//To prevent tampering with the original values
+            MinecraftForge.EVENT_BUS.post(new VillagerTradePopulationEvent(this, trades, skillLevel));//Allows modders to modify the trade list.
+            return trades.toArray(new ITradeList[]{});
```
I'm not convinced that there needs to be a special event for this to make each instance editable.
ANd if there is this is NOT the place to put it as this doesn't have ANY context beyond the global career info.
You'd want to edit the per-entity list not the global.

If you WANT to edit the global then you'd do something like Career c = registry("farmer").level(1); c.remove(trade); c.add(trade)

thilokru commented:
Thanks for the comments. I'll work on it.

thilokru commented:
This should fix the stuff you complained about.

Willie: I added documentation to my Event

Lex:
- I just left the code adding the vanilla trades. It works.
- I did not reformat this time, so that there is no "reformatting spam" cluttering my PR
- No bouncer methods
- Switched trades to an object oriented system, where the trades are editable via the VillagerCareers
- Wrote code, so that after all a two-dimensional array is returned to the EntityVillager. I left the patch untouched except for the line which is responsible for the manipulation of the trade list. This is where the event is fired now.

I hope that there are no further errors. Otherwise, please let me know.

williewillus commented on src/main/java/net/minecraftforge/event/entity/living/VillagerTradePopulationEvent.java:
```diff
@@ -0,0 +1,64 @@
+ * The VillagerTradePopulationEvent is fired when the trades are populated<br>
+ * in {@link EntityVillager#populateBuyingList()}.<br>
+ * It allows the modder to modify the trades of an individual Villager.<br>
+ * This event is fired via the {@link ForgeHooks#onVillagerTradePopulation}.<br>
```
Don't put ForgeHooks, we know everything is fired from there.
Put where in vanilla was patched to call Forgehooks
This also helps us make sure it's properly updated every MC Version

Something like "This event is fired from ForgeHooks... via <vanilla method>"

thilokru replied:
I think I mentioned it with:
> The VillagerTradePopulationEvent is fired when the trades are populated in {@link EntityVillager#populateBuyingList()}
This says, that it is called from EntityVillager.populateBuyingList()

williewillus replied:
Oh I did not see that, derp
Carry on :P

ganymedes01 commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -212,6 +214,15 @@ public VillagerCareer getCareer(int id)
+        {
+            for (VillagerCareer car : this.careers)
+            {
+                if (car.getName() == name)
```
Always use ```equals``` when comparing Strings in Java

thilokru replied:
Yeah, missed that one. Gonna be fixed.

FRedEnergy commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -970,4 +975,11 @@ public static void onEmptyClick(EntityPlayer player, EnumHand hand)
+    {
+        VillagerTradePopulationEvent event = new VillagerTradePopulationEvent(villager, villager.getProfessionForge().getCareer(forgeCareerID), Lists.newArrayList(originalTrades));
+        MinecraftForge.EVENT_BUS.post(event);
+        return event.getTrades().toArray(new ITradeList[]{});
```
Do not call `toArray` with zero length array because it will cause array allocation twice
For the perfomance improvements you can do it like this
```
event.getTrades().toArray(new ITradeList[event.getTrades().size()])
```

diesieben07 replied:
Making this blanket statement ("X is faster") is not correct. See [this investigation](http://shipilev.net/blog/2016/arrays-wisdom-ancients/).

FRedEnergy commented on src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java:
```diff
@@ -224,30 +235,66 @@ public int getRandomCareer(Random rand)
+            ITradeList[][] returnValue = new ITradeList[this.trades.size()][];
+            for(int i = 0; i < this.trades.size(); i++)
+            {
+               returnValue[i] = this.trades.get(i).toArray(new ITradeList[]{});
```
Same thing, don't do it like this.
Instead do:
`trades.get(i).toArray(new ITradeList[trades.get(i).size()])`
Yo also can move the result of `trades.get(i)` into a local variable
```
List<ITradeList> trades = this.trades.get(i);
returnValue[i] = trades.toArray(new ITradeList[trades.size()]);
```

thilokru commented:
This PR should be done now. I'd like to ask again if there are any further problems which would prevent my branch from being merged.

thilokru commented:
Rebased my commits to no longer conflict with the latest changes.

Actuarius commented:
@williewillus added labels [Feature, Bug]

thilokru commented:
As part of my code has been implemented independently by Lex in one of the latest commits, I'm going to open a new PR with the remaining features. Therefore this PR will be closed soon.
